account_invoice_change_currency/models/account_move.py

```diff
@@ -0,0 +1,93 @@
+from odoo import api, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    custom_rate = fields.Float(
+        digits=(12, 6),
+        default=1,
+        store=True,
+        readonly=False,
+        compute="_compute_currency_change_rate",
+        help="Set new currency rate to apply on the invoice.\n"
+        "This rate will be taken in order to convert amounts between the "
+        "currency on the invoice and last currency",
+    )
+    created_invoice_currency_id = fields.Many2one(
+        "res.currency",
+        help="Store the original currency when the invoice was created. "
+        "This is used to calculate conversion from this currency",
+    )
+    is_original_currency = fields.Boolean(
+        compute="_compute_is_original_currency",
+        help="Check if current currency is the original currency. "
+        "This is used to hide custom rate field in the form view",
+    )
+
+    @api.model
+    def create(self, values):
+        values.setdefault("created_invoice_currency_id", values.get("currency_id"))
+        return super().create(values)
+
+    def action_account_change_currency(self):
+        """
+        This method convert the original price unit from the original
+        currency when the invoice was created to the current currency
+        using the custom rate and recompute all taxes.
+        """
+        today = fields.Date.context_today(self)
+        invoices = self.filtered(lambda x: x.state == "draft").with_context(
+            check_move_validity=False,
+        )
+        for invoice in invoices:
+            invoice_date = invoice.invoice_date or today
+            to_currency = invoice.currency_id
+            context = {"custom_rate": invoice.custom_rate, "to_currency": to_currency}
+            original_currency = invoice.created_invoice_currency_id.with_context(
+                **context
+            )
+            for line in invoice.line_ids:
+                line.price_unit = original_currency._convert(
+                    line.original_price_unit,
+                    to_currency,
+                    invoice.company_id,
+                    invoice_date,
+                )
+            invoice._recompute_dynamic_lines(recompute_all_taxes=True)
+
+    @api.depends("company_id", "currency_id", "invoice_date")
+    def _compute_currency_change_rate(self):
+        """
+        Compute the custom rate from the original currency when the invoice
+        was created to the current currency. The custom rate field is
+        editable, but it will not change if custom rate is zero or the
+        current currency and the original currency are the same
+        """
+        for invoice in self:
+            date = invoice.invoice_date or fields.Date.context_today(invoice)
+            from_currency = (
+                invoice.created_invoice_currency_id or invoice.company_currency_id
+            )
+            if not invoice.currency_id:
```

Comment:
Compute methods should always set a value. If any value is missing (including company), set it to 1.0. The if should be the first thing inside the loop

Reply:
Done!